managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2718,6 +2719,72 @@ void setReadPosition(Position newReadPositionInt) {
+     * - This method is valid regardless of the consumer ACK type.
+     * - If there is a consumer ack request after this event, it will also work.
+     */
+    @Override
+    public void noticeNonRecoverableLedgerSkipped(final long ledgerId){
+        LedgerInfo ledgerInfo = ledger.getLedgersInfo().get(ledgerId);
+        if (ledgerInfo == null) {
+            return;
+        }
+        lock.writeLock().lock();
+        log.warn("[{}] [{}] Since the ledger [{}] is lost and the autoSkipNonRecoverableData is true, this ledger will"
+                + " be auto acknowledge in subscription", ledger.getName(), name, ledgerId);
+        try {
+            List<Position> positionsToAck = new ArrayList<>();
+            for (int i = 0; i < ledgerInfo.getEntries(); i++) {
+                if (!individualDeletedMessages.contains(ledgerId, i)) {
+                    positionsToAck.add(PositionImpl.get(ledgerId, i));
+                }
+                // Acknowledge in segments to avoid OOM.
+                if (positionsToAck.size() >= 1000) {
+                    retryToAcknowledgeNonRecoverablePositions(positionsToAck, 1);
+                    positionsToAck = new ArrayList<>();
+                }
+            }
+            // Acknowledge the last segments.
+            if (!positionsToAck.isEmpty()) {
+                retryToAcknowledgeNonRecoverablePositions(positionsToAck, 1);
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private void retryToAcknowledgeNonRecoverablePositions(List<Position> positions, int retryTimes) {
+        if (CollectionUtils.isEmpty(positions)) {
+            return;
+        }
+        asyncDelete(positions, new AsyncCallbacks.DeleteCallback() {
+            @Override
+            public void deleteComplete(Object ctx) {
+                // ignore.
+            }
+
+            @Override
+            public void deleteFailed(ManagedLedgerException ex, Object ctx) {
+                if (retryTimes <= 3) {
+                    log.warn("[{}] [{}] Try to acknowledge the non recoverable positions fail and it will be retry"
+                                    + " after 60s. ledgerId: {}, the current retry times: {}",
+                            ledger.getName(), name, positions.get(0).getLedgerId(), retryTimes, ex);
+                    ledger.getScheduledExecutor()
+                            .schedule(() -> retryToAcknowledgeNonRecoverablePositions(positions, retryTimes + 1),
+                                    60, TimeUnit.SECONDS);
+                } else {
+                    log.error("[{}] [{}] Try to acknowledge the non recoverable positions ultimately failed."
+                                    + " ledgerId: {}, retry times: {}",
+                            ledger.getName(), name, positions.get(0).getLedgerId(), retryTimes, ex);
+                }
+            }
+        }, retryTimes);
```

Comment:
The method `internalMarkDelete` already handled the failure operation, why do we need to handle it again?
We only need to make sure the memory state is updated. The failure of ack state persistence is fine, right? Because the subsequent ack operations will try again.

If the broker crashed, the non-recoverable ledger will be detected again?

Reply:
fixed